<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Data preprocessing &mdash; TandemMod 1.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=fc837d61"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            TandemMod
          </a>
              <div class="version">
                1.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing.html">Data preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../run_examples.html">Run examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_postprocessing.html">Data postprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../help.html">Getting Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citing TandemMod</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TandemMod</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Data preprocessing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/.ipynb_checkpoints/data_preprocessing-checkpoint.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="data-preprocessing">
<span id="id1"></span><h1>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Link to this heading"></a></h1>
<p>This section involves transforming the raw electrical signal data from nanopore direct RNA sequencing (DRS) into meaningful features that can be used to study RNA modifications. The preprocessing procedure typically consists of several stages.</p>
<ul class="simple">
<li><p>Raw FAST5 files: Begin with the raw FAST5 files generated by the nanopore sequencing platform. These files contain the raw electrical signal data captured during sequencing.</p></li>
<li><p>Basecalling: Perform basecalling on the raw FAST5 files using basecalling software such as Guppy or Albacore. Basecalling converts the raw electrical signals into sequences (e.g., A, C, G, T).</p></li>
<li><p>Resquiggling: Extract events from the basecalled FAST5 files. Events represent segments of the raw signal associated with individual bases. This step involves aligning the corrected sequences with the raw signal data.</p></li>
<li><p>Feature extraction: Extract features from the events to represent various characteristics of the RNA modifications. Features can include event duration, mean, standard error and other signal-based properties.</p></li>
<li><p>Preprocessing and normalization: Preprocess the extracted features by applying normalization techniques such as scaling or log transformation. Normalization helps to remove any biases or variations in the features across different samples.</p></li>
<li><p>Dataset creation: Combine the preprocessed features to create a dataset, which will be used for training or prediction purposes.</p></li>
<li><p>Additional processing: Depending on the specific requirements of the analysis, additional processing steps may be performed, such as filtering out low-quality events, removing noise, or applying statistical methods.</p></li>
</ul>
<section id="basecalling">
<h2>Basecalling<a class="headerlink" href="#basecalling" title="Link to this heading"></a></h2>
<p>Guppy is used for basecalling in TandemMod. Guppy, as well as the now deprecated Albacore and all other basecallers, uses files in fast5 format as input. In addition to basecalling, Guppy also performs filtering of low quality reads, clipping of Oxford Nanopore adapters. More detailed documentation about Guppy can be found on the official <a class="reference external" href="https://github.com/nanoporetech/pyguppyclient">Nanopore Technology repository</a>. This step can be time-consuming and may require several hours or even days to complete, depending on the computational capacity available:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">guppy_basecaller</span> <span class="o">-</span><span class="n">i</span> <span class="n">demo</span><span class="o">/</span><span class="n">fast5</span> <span class="o">-</span><span class="n">s</span> <span class="n">demo</span><span class="o">/</span><span class="n">guppy</span> <span class="o">--</span><span class="n">num_callers</span> <span class="mi">40</span> <span class="o">--</span><span class="n">recursive</span> <span class="o">--</span><span class="n">fast5_out</span> <span class="o">--</span><span class="n">config</span> <span class="n">rna_r9</span><span class="mf">.4.1_70</span><span class="n">bps_hac</span><span class="o">.</span><span class="n">cfg</span>
</pre></div>
</div>
<p>Arguments</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument name</p></th>
<th class="head"><p>Required</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>-i=DIR</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Input directory, containing FAST5 files generated by the nanopore sequencing platform</p></td>
</tr>
<tr class="row-odd"><td><p>-s=DIR</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Output directory, containing FAST5 file as well as basecalled sequences.</p></td>
</tr>
<tr class="row-even"><td><p>–num_callers=NUM</p></td>
<td><p>No</p></td>
<td><p>1</p></td>
<td><p>Number of processes to run.</p></td>
</tr>
<tr class="row-odd"><td><p>–fast5_out</p></td>
<td><p>Yes</p></td>
<td><p>None</p></td>
<td><p>Output FAST5 files to the directory.</p></td>
</tr>
<tr class="row-even"><td><p>–config=STR</p></td>
<td><p>Yes</p></td>
<td><p>None</p></td>
<td><p>The configure file is “rna_r9.4.1_70bps_hac.cfg” in TandemMod and should be adjusted according to DRS platform.</p></td>
</tr>
<tr class="row-odd"><td><p>–recursive</p></td>
<td><p>Yes</p></td>
<td><p>None</p></td>
<td><p>This Argument allows recursive processing or batch processing of files</p></td>
</tr>
</tbody>
</table>
</section>
<section id="multi-fast5-to-single-fast5">
<h2>Multi-fast5 to single-fast5<a class="headerlink" href="#multi-fast5-to-single-fast5" title="Link to this heading"></a></h2>
<p>If fast5 reads are stored at multi-reads format, ont_fast5_api is recommended to convert multi-fast5 reads to single-fast5 reads. Usually, the size of multi-reads fast5 file is about 200-300M. Convert multi-reads files to single-read files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">multi_to_single_fast5</span> <span class="o">-</span><span class="n">i</span> <span class="n">demo</span><span class="o">/</span><span class="n">guppy</span> <span class="o">-</span><span class="n">s</span> <span class="n">demo</span><span class="o">/</span><span class="n">guppy_single</span> <span class="o">-</span><span class="n">t</span> <span class="mi">40</span> <span class="o">--</span><span class="n">recursive</span>
</pre></div>
</div>
<p>Arguments</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument name</p></th>
<th class="head"><p>Required</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>-i=DIR</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Input directory, containing multi-reads FAST5 files.</p></td>
</tr>
<tr class="row-odd"><td><p>-s=DIR</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Output directory, containing single-read FAST5 files.</p></td>
</tr>
<tr class="row-even"><td><p>-t=NUM</p></td>
<td><p>No</p></td>
<td><p>1</p></td>
<td><p>Number of processes to run.</p></td>
</tr>
<tr class="row-odd"><td><p>–recursive</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>This Argument allows recursive processing or batch processing of files.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="resquiggling">
<h2>Resquiggling<a class="headerlink" href="#resquiggling" title="Link to this heading"></a></h2>
<p>The resquiggling algorithm is the basis for the Tombo framework. It takes as input a read file (in FAST5 format) containing raw signal and associated base calls. The base calls are mapped to a genome or transcriptome reference and then the raw signal is assigned to the reference sequence based on an expected current level model. Tombo is used for resquiggling in TandemMod:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tombo</span> <span class="n">resquiggle</span> <span class="o">--</span><span class="n">overwrite</span> <span class="o">--</span><span class="n">basecall</span><span class="o">-</span><span class="n">group</span> <span class="n">Basecall_1D_001</span> <span class="n">demo</span><span class="o">/</span><span class="n">guppy_single</span>  <span class="n">demo</span><span class="o">/</span><span class="n">reference_transcripts</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">processes</span> <span class="mi">40</span> <span class="o">--</span><span class="n">fit</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="n">scale</span> <span class="o">--</span><span class="n">include</span><span class="o">-</span><span class="n">event</span><span class="o">-</span><span class="n">stdev</span>
</pre></div>
</div>
<p>Arguments</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument name</p></th>
<th class="head"><p>Required</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>–overwrite</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Overwrite previous corrected group in FAST5 files.</p></td>
</tr>
<tr class="row-odd"><td><p>–basecall-group</p></td>
<td><p>No</p></td>
<td><p>Basecall_1D_000</p></td>
<td><p>FAST5 group obtain original basecalls.</p></td>
</tr>
<tr class="row-even"><td><p>–processes</p></td>
<td><p>No</p></td>
<td><p>1</p></td>
<td><p>Number of processes to run.</p></td>
</tr>
<tr class="row-odd"><td><p>–fit-global-scale</p></td>
<td><p>No</p></td>
<td><p>NA</p></td>
<td><p>Apply a scaling factor.</p></td>
</tr>
<tr class="row-even"><td><p>–include-event-stdev</p></td>
<td><p>No</p></td>
<td><p>NA</p></td>
<td><p>Include the standard deviation.</p></td>
</tr>
<tr class="row-odd"><td><p>args[0]</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Fast5 basedir.</p></td>
</tr>
<tr class="row-even"><td><p>args[1]</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Reference transcripts, in fasta format.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="feature-extraction">
<h2>Feature extraction<a class="headerlink" href="#feature-extraction" title="Link to this heading"></a></h2>
<p>minimap2 is used to map basecalled sequences to reference transcripts:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="n">demo</span><span class="o">/</span><span class="n">guppy</span><span class="o">/</span><span class="k">pass</span><span class="o">/*.</span><span class="n">fastq</span> <span class="o">&gt;</span><span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">minimap2</span> <span class="o">-</span><span class="n">ax</span> <span class="nb">map</span><span class="o">-</span><span class="n">ont</span> <span class="n">demo</span><span class="o">/</span><span class="n">reference_transcripts</span><span class="o">.</span><span class="n">fasta</span> <span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">fastq</span> <span class="o">&gt;</span><span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">sam</span>
</pre></div>
</div>
<p>Extract signal files from FAST5 files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">extract_signal_from_fast5</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">p</span> <span class="mi">40</span> <span class="o">--</span><span class="n">fast5</span> <span class="n">demo</span><span class="o">/</span><span class="n">guppy_single</span> <span class="o">--</span><span class="n">reference</span> <span class="n">demo</span><span class="o">/</span><span class="n">reference_transcripts</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">sam</span> <span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">sam</span> <span class="o">--</span><span class="n">output</span> <span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">signal</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">clip</span> <span class="mi">10</span>
</pre></div>
</div>
<p>Arguments</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument name</p></th>
<th class="head"><p>Required</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>–fast5</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Fast5 basedir.</p></td>
</tr>
<tr class="row-odd"><td><p>–reference</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Reference transcripts, in fasta format.</p></td>
</tr>
<tr class="row-even"><td><p>-p</p></td>
<td><p>No</p></td>
<td><p>1</p></td>
<td><p>Number of processes to run.</p></td>
</tr>
<tr class="row-odd"><td><p>–sam</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Aligment results, output from minimap2.</p></td>
</tr>
<tr class="row-even"><td><p>–output</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Output file contraining current signals.</p></td>
</tr>
<tr class="row-odd"><td><p>–clip</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Base clip at both ends.</p></td>
</tr>
</tbody>
</table>
<p>Extract features from signal files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">extract_feature_from_signal</span><span class="o">.</span><span class="n">py</span>  <span class="o">--</span><span class="n">signal_file</span> <span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">signal</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">clip</span> <span class="mi">10</span> <span class="o">--</span><span class="n">output</span> <span class="n">demo</span><span class="o">/</span><span class="n">m6A</span><span class="o">.</span><span class="n">feature</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">motif</span> <span class="n">DRACH</span>
</pre></div>
</div>
<p>Arguments</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument name</p></th>
<th class="head"><p>Required</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>–signal_file</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>File contraining current signals.</p></td>
</tr>
<tr class="row-odd"><td><p>–reference</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Reference transcripts, in fasta format.</p></td>
</tr>
<tr class="row-even"><td><p>–output</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Output file contraining features.</p></td>
</tr>
<tr class="row-odd"><td><p>–clip</p></td>
<td><p>Yes</p></td>
<td><p>NA</p></td>
<td><p>Base clip at both ends.</p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Yulab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>