<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Installation &mdash; TandemMod 1.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=fc837d61"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            TandemMod
          </a>
              <div class="version">
                1.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing.html">Data preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../run_examples.html">Run examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_postprocessing.html">Data postprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../help.html">Getting Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citing TandemMod</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TandemMod</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Installation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/.ipynb_checkpoints/installation-checkpoint.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="installation">
<span id="id1"></span><h1>Installation<a class="headerlink" href="#installation" title="Link to this heading"></a></h1>
<p>The following modules are needed to run TandemMod.</p>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Required modules</span><a class="headerlink" href="#id2" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>module</p></th>
<th class="head"><p>version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>minimap2</p></td>
<td><p>2.17-r941</p></td>
</tr>
<tr class="row-odd"><td><p>python</p></td>
<td><p>3.7.12</p></td>
</tr>
<tr class="row-even"><td><p>h5py</p></td>
<td><p>3.7.0</p></td>
</tr>
<tr class="row-odd"><td><p>statsmodels</p></td>
<td><p>0.10.0</p></td>
</tr>
<tr class="row-even"><td><p>joblib</p></td>
<td><p>0.16.0</p></td>
</tr>
<tr class="row-odd"><td><p>scikit-learn</p></td>
<td><p>0.22</p></td>
</tr>
<tr class="row-even"><td><p>torch</p></td>
<td><p>1.9.1</p></td>
</tr>
<tr class="row-odd"><td><p>guppy</p></td>
<td><p>6.1.5</p></td>
</tr>
<tr class="row-even"><td><p>ont-tombo</p></td>
<td><p>1.5.1</p></td>
</tr>
<tr class="row-odd"><td><p>ont_vbz_hdf_plugin</p></td>
<td><p>1.0.1</p></td>
</tr>
<tr class="row-even"><td><p>ont-fast5-api</p></td>
<td><p>4.1.1</p></td>
</tr>
<tr class="row-odd"><td><p>numpy</p></td>
<td><p>1.19.5</p></td>
</tr>
</tbody>
</table>
<p>Conda is recommended for package management, you can create a new conda environment and then install the packages. Here’s an example of how you can do it. Create a new conda environment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">-</span><span class="n">n</span> <span class="n">TandemMod</span> <span class="n">python</span><span class="o">=</span><span class="mf">3.7.12</span>
</pre></div>
</div>
<p>Activate the newly created environment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">activate</span> <span class="n">TandemMod</span>
</pre></div>
</div>
<p>Install the required modules:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span>
<span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">bioconda</span>

<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">scipy</span><span class="o">=</span><span class="mf">1.7.0</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">minimap2</span><span class="o">=</span><span class="mf">2.17</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">numpy</span><span class="o">=</span><span class="mf">1.19.5</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">anaconda</span> <span class="n">h5py</span><span class="o">=</span><span class="mf">3.7.0</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">joblib</span><span class="o">=</span><span class="mf">0.16.0</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">anaconda</span> <span class="n">scikit</span><span class="o">-</span><span class="n">learn</span><span class="o">=</span><span class="mf">0.22</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">ont</span><span class="o">-</span><span class="n">tombo</span><span class="o">=</span><span class="mf">1.5.1</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">ont_vbz_hdf_plugin</span><span class="o">=</span><span class="mf">1.0.1</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">ont</span><span class="o">-</span><span class="n">fast5</span><span class="o">-</span><span class="n">api</span><span class="o">=</span><span class="mf">4.1.1</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">statsmodels</span><span class="o">=</span><span class="mf">0.10.0</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">torch</span><span class="o">==</span><span class="mf">1.9.1</span>
</pre></div>
</div>
<p>Or, some of the modules can be installed by pip:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">numpy</span><span class="o">==</span><span class="mf">1.19.5</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">h5py</span><span class="o">==</span><span class="mf">3.7.0</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">statsmodels</span><span class="o">==</span><span class="mf">0.10.0</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">joblib</span><span class="o">==</span><span class="mf">0.16.0</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">scikit</span><span class="o">-</span><span class="n">learn</span><span class="o">==</span><span class="mf">0.22</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">ont</span><span class="o">-</span><span class="n">tombo</span><span class="o">==</span><span class="mf">1.5.1</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">ont</span><span class="o">-</span><span class="n">fast5</span><span class="o">-</span><span class="n">api</span><span class="o">==</span><span class="mf">4.1.1</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">scipy</span><span class="o">==</span><span class="mf">1.7.0</span>
</pre></div>
</div>
<p>Guppy can be obtained from <a class="reference external" href="https://nanoporetech.com/">Oxford Nanopore Technologies</a> or from this <a class="reference external" href="https://mirror.oxfordnanoportal.com/software/analysis/ont-guppy-cpu-6.1.5-1.el7.x86_64.rpm">mirror</a>. Install Guppy using dpkg:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">alien</span> <span class="n">ont</span><span class="o">-</span><span class="n">guppy</span><span class="o">-</span><span class="n">cpu</span><span class="o">-</span><span class="mf">6.1.5</span><span class="o">-</span><span class="mf">1.</span><span class="n">el7</span><span class="o">.</span><span class="n">x86_64</span><span class="o">.</span><span class="n">rpm</span>
<span class="n">dpkg</span> <span class="o">-</span><span class="n">i</span> <span class="n">ont</span><span class="o">-</span><span class="n">guppy</span><span class="o">-</span><span class="n">cpu</span><span class="o">-</span><span class="mf">6.1.5</span><span class="o">-</span><span class="mf">1.</span><span class="n">el7</span><span class="o">.</span><span class="n">x86_64</span><span class="o">.</span><span class="n">deb</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">libhdf5</span></code> and <code class="docutils literal notranslate"><span class="pre">libcrypto</span></code> are required for running guppy.</p>
<p>After installing all the essential packages,  reset the environment’s state by deactivating and reactivating the environment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">deactivate</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">TandemMod</span>
</pre></div>
</div>
<p>We have also provided a yaml file in the repository so you can install the dependencies through the configuration file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">TandemMod</span><span class="o">.</span><span class="n">yaml</span>
</pre></div>
</div>
<p>The source code and data processing scripts are available on <a class="reference external" href="https://github.com/yulab2021/TandemMod">GitHub</a>. You can download them by using the git clone command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">yulab2021</span><span class="o">/</span><span class="n">TandemMod</span><span class="o">.</span><span class="n">git</span>
</pre></div>
</div>
<p>TandemMod offers three modes: de novo training, transfer learning, and prediction. Researchers can train from scratch, fine-tune pre-trained models, or apply existing models for predictions. It provides a user-friendly solution for studying RNA modifications.</p>
<p>In the provided repository, the pretrained models are located under the <code class="docutils literal notranslate"><span class="pre">./models</span></code> directory, and the data processing scripts and the main script are located under the <code class="docutils literal notranslate"><span class="pre">./scripts</span></code> directory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>.
├── data
│   ├── A_test.tsv
│   ├── A_train.tsv
│   ├── m5C
│   ├── m6A
│   ├── m6A_test.tsv
│   └── m6A_train.tsv
├── demo
│   ├── fast5
│   │   └── batch_0.fast5
│   ├── files.txt
│   ├── guppy
│   │   ├── fail
│   │   │   └── fastq_runid_71d544d3bd9e1fe7886a5d176c756a576d30ed50_0_0.fastq
│   │   ├── guppy_basecaller_log-2023-06-06_09-58-28.log
│   │   ├── pass
│   │   │   └── fastq_runid_71d544d3bd9e1fe7886a5d176c756a576d30ed50_0_0.fastq
│   │   ├── sequencing_summary.txt
│   │   ├── sequencing_telemetry.js
│   │   └── workspace
│   │       └── batch_0.fast5
├── models
│   ├── hm5C_transfered_from_m5C.pkl
│   ├── m1A_train_on_rice_cDNA.pkl
│   ├── m5C_train_on_rice_cDNA.pkl
│   ├── m6A_train_on_rice_cDNA.pkl
│   ├── m7G_transfered_from_m5C.pkl
│   ├── psU_transfered_from_m5C.pkl
│   ├── test.model
│   └── test.pkl
├── plot
├── README.md
├── scripts
│   ├── extract_feature_from_signal.py
│   ├── extract_signal_from_fast5.py
│   ├── __init__.py
│   ├── models.py
│   ├── TandemMod.py
│   ├── train_test_split.py
│   ├── transcriptome_loci_to_genome_loci.py
│   └── utils.py
└── TandemMod.yaml
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Yulab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>